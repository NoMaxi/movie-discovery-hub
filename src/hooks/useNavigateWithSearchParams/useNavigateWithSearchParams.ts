import { useCallback } from "react";
import { useNavigate, useSearchParams } from "react-router-dom";

export const useNavigateWithSearchParams = () => {
    const navigate = useNavigate();
    const [searchParams] = useSearchParams();

    return useCallback(
        (path: string) => {
            const searchParamsString = searchParams.toString();
            navigate(`${path}${searchParamsString ? `?${searchParamsString}` : ""}`);
        },
        [navigate, searchParams],
    );
};
