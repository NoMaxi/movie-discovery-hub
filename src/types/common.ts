export type Genre = "All" | "Documentary" | "Comedy" | "Horror" | "Crime";

export interface Movie {
    id: string;
    imageUrl: string;
    title: string;
    releaseYear: number;
    genres: string[];
}
