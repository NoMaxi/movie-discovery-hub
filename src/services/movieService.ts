import axios, { AxiosResponse } from "axios";
import { GenreFilter, Movie, SortOption } from "@/types/common";
import { DEFAULT_MOVIES_PER_PAGE } from "@/constants/constants";
import { mapAPIMovieDetailsToMovie, mapSortOptionToSortBy, mapSortOptionToSortOrder } from "@/utils/movieMapping";

export interface CreateMoviePayload {
    genres: string[];
    overview: string;
    poster_path: string;
    runtime: number;
    title: string;
    budget?: number;
    release_date?: string;
    revenue?: number;
    tagline?: string;
    vote_average?: number;
    vote_count?: number;
}

export interface APIMovieDetails extends Omit<CreateMoviePayload, "runtime"> {
    id: number;
    runtime: number | null;
}

export interface MoviesResponse {
    data: APIMovieDetails[];
    totalAmount: number;
    offset: number;
    limit: number;
}

export type SearchBy = "title" | "genres";
export type SortBy = "title" | "release_date";
export type SortOrder = "asc" | "desc";

export interface MoviesRequestParams {
    searchBy?: SearchBy;
    sortBy?: SortBy;
    sortOrder?: SortOrder;
    search?: string;
    limit?: number;
    offset?: number;
    filter?: GenreFilter;
}

export interface InfiniteMovieListResult {
    movies: Movie[];
    totalAmount: number;
    offset: number;
    limit: number;
    nextOffset?: number;
}

export const API_URL = "http://localhost:4000/movies";

export const movieService = {
    async getMovies(
        searchQuery: string,
        sortCriterion: SortOption,
        activeGenre: GenreFilter,
        signal: AbortSignal | undefined,
        offset: number = 0,
    ): Promise<InfiniteMovieListResult> {
        const params: MoviesRequestParams = {
            sortBy: mapSortOptionToSortBy(sortCriterion),
            sortOrder: mapSortOptionToSortOrder(sortCriterion),
            searchBy: "title",
            limit: DEFAULT_MOVIES_PER_PAGE,
            offset,
        };

        if (searchQuery) {
            params.search = searchQuery;
        }

        if (activeGenre !== "All") {
            params.filter = activeGenre;
        }

        const { data: moviesResponse } = await axios.get<MoviesResponse>(API_URL, { params, signal });

        const { offset: responseOffset, totalAmount, limit, data } = moviesResponse;
        const movies = data.map(mapAPIMovieDetailsToMovie);
        const currentTotalFetched = offset + movies.length;
        const nextOffset = currentTotalFetched < totalAmount ? currentTotalFetched : undefined;

        return {
            movies,
            totalAmount,
            limit,
            offset: responseOffset,
            nextOffset,
        };
    },

    async getAPIMovieDetailsById(id: number): Promise<APIMovieDetails> {
        const response = await axios.get<APIMovieDetails>(`${API_URL}/${id}`);
        return response.data;
    },

    async getMovieById(id: number): Promise<Movie> {
        const movieDetails = await this.getAPIMovieDetailsById(id);
        return mapAPIMovieDetailsToMovie(movieDetails);
    },

    async createMovie(movieData: CreateMoviePayload): Promise<Movie> {
        const { data } = await axios.post<APIMovieDetails, AxiosResponse<APIMovieDetails>, CreateMoviePayload>(
            API_URL,
            movieData,
        );
        return mapAPIMovieDetailsToMovie(data);
    },

    async updateMovie(movieData: APIMovieDetails): Promise<Movie> {
        const { data } = await axios.put<APIMovieDetails, AxiosResponse<APIMovieDetails>, APIMovieDetails>(
            API_URL,
            movieData,
        );
        return mapAPIMovieDetailsToMovie(data);
    },

    async deleteMovie(id: number): Promise<void> {
        await axios.delete(`${API_URL}/${id}`);
    },
};
