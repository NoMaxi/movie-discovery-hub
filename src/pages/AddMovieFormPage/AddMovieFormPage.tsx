import { useState } from "react";
import { movieService } from "@/services/movieService";
import { MovieFormData } from "@/types/common";
import { useNavigateWithSearchParams } from "@/hooks/useNavigateWithSearchParams/useNavigateWithSearchParams";
import { ErrorMessage } from "@/components/common/ErrorMessage/ErrorMessage";
import { Dialog } from "@/components/common/Dialog/Dialog";
import { MovieForm } from "@/components/MovieForm/MovieForm";

export const AddMovieFormPage = () => {
    const navigateWithSearchParams = useNavigateWithSearchParams();
    const [isSubmitting, setIsSubmitting] = useState(false);
    const [errorMessage, setErrorMessage] = useState<string | null>(null);

    const handleSubmit = async (formData: MovieFormData) => {
        setErrorMessage(null);
        setIsSubmitting(true);

        try {
            const movie = await movieService.createMovie(formData);
            window.scrollTo({ top: 0, behavior: "smooth" });
            navigateWithSearchParams(`/${movie.id}`);
        } catch (error) {
            console.error("Error creating a movie:", error);
            setErrorMessage("Failed to create a movie. Please try again.");
        } finally {
            setIsSubmitting(false);
        }
    };

    const handleClose = () => {
        navigateWithSearchParams("/");
    };

    const showErrorMessage = !isSubmitting && !!errorMessage;

    return (
        <Dialog title="Add Movie" onClose={handleClose} className="max-w-[976px]">
            {showErrorMessage && (
                <ErrorMessage message={errorMessage} onDismiss={() => setErrorMessage(null)} className="pb-6" />
            )}

            <MovieForm onSubmit={handleSubmit} isLoading={isSubmitting} resetMode="clear" />
        </Dialog>
    );
};
