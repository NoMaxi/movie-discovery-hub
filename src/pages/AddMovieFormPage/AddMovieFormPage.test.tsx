import { ReactNode } from "react";
import { render, screen, waitFor, act } from "@testing-library/react";
import "@testing-library/jest-dom";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { movieService } from "@/services/movieService";
import { useNavigateWithSearchParams } from "@/hooks/useNavigateWithSearchParams/useNavigateWithSearchParams";
import { AddMovieFormPage } from "./AddMovieFormPage";
import { MovieFormData } from "@/types/common";

jest.mock("@/services/movieService");
jest.mock("@/hooks/useNavigateWithSearchParams/useNavigateWithSearchParams");

let mockDialogOnClose: () => void;
jest.mock("@/components/common/Dialog/Dialog", () => ({
    Dialog: ({
        title,
        children,
        onClose,
        className,
    }: {
        title: string;
        children: ReactNode;
        onClose: () => void;
        className?: string;
    }) => {
        mockDialogOnClose = onClose;
        return (
            <div data-testid="mock-dialog" className={className}>
                <h2 data-testid="mock-dialog-title">{title}</h2>
                <button data-testid="mock-dialog-close-button" onClick={onClose}>
                    Close
                </button>
                {children}
            </div>
        );
    },
}));

let mockMovieFormSubmit: (formData: MovieFormData) => void;
let lastMovieFormProps: { isLoading?: boolean; resetMode?: string } | null = null;

jest.mock("@/components/MovieForm/MovieForm", () => ({
    MovieForm: ({
        onSubmit,
        isLoading,
        resetMode,
    }: {
        onSubmit: (formData: MovieFormData) => void;
        isLoading?: boolean;
        resetMode?: string;
    }) => {
        mockMovieFormSubmit = onSubmit;
        lastMovieFormProps = { isLoading, resetMode };
        return <form data-testid="mock-movie-form">Mock Movie Form</form>;
    },
}));

let mockErrorMessageDismiss: (() => void) | undefined;
jest.mock("@/components/common/ErrorMessage/ErrorMessage", () => ({
    ErrorMessage: ({
        message,
        onDismiss,
        className,
    }: {
        message: string | null;
        onDismiss: () => void;
        className?: string;
    }) => {
        mockErrorMessageDismiss = onDismiss;
        return message ? (
            <div data-testid="mock-error-message" className={className}>
                {message}
                <button data-testid="mock-error-dismiss-button" onClick={onDismiss}>
                    Dismiss
                </button>
            </div>
        ) : null;
    },
}));

const mockNavigateWithSearchParams = jest.fn();
const mockCreateMovie = movieService.createMovie as jest.Mock;

const renderComponent = () => {
    const queryClient = new QueryClient();
    return render(
        <QueryClientProvider client={queryClient}>
            <AddMovieFormPage />
        </QueryClientProvider>,
    );
};

const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

describe("AddMovieFormPage", () => {
    beforeEach(() => {
        jest.clearAllMocks();
        (useNavigateWithSearchParams as jest.Mock).mockReturnValue(mockNavigateWithSearchParams);
        window.scrollTo = jest.fn();
        mockErrorMessageDismiss = undefined;
        lastMovieFormProps = null;
    });

    test("Should render correctly with Dialog and MovieForm", () => {
        const { asFragment } = renderComponent();

        expect(screen.getByTestId("mock-dialog")).toBeInTheDocument();
        expect(screen.getByTestId("mock-dialog-title")).toHaveTextContent("Add Movie");
        expect(screen.getByTestId("mock-movie-form")).toBeInTheDocument();
        expect(lastMovieFormProps?.resetMode).toBe("clear");
        expect(lastMovieFormProps?.isLoading).toBe(false);
        expect(screen.queryByTestId("mock-error-message")).not.toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should handle successful form submission", async () => {
        const mockFormData: MovieFormData = {
            title: "New Movie",
            poster_path: "path/to/image.jpg",
            genres: ["Action"],
            overview: "A great movie",
            runtime: 120,
            release_date: "2023-01-01",
            vote_average: 8.0,
        };
        const mockCreatedMovie = { ...mockFormData, id: 123 };

        mockCreateMovie.mockImplementation(async () => {
            await delay(10);
            return mockCreatedMovie;
        });

        renderComponent();

        await act(async () => {
            mockMovieFormSubmit(mockFormData);
        });

        await waitFor(() => expect(lastMovieFormProps?.isLoading).toBe(true));

        expect(mockCreateMovie).toHaveBeenCalledTimes(1);
        expect(mockCreateMovie).toHaveBeenCalledWith(mockFormData);

        await waitFor(() => {
            expect(window.scrollTo).toHaveBeenCalledWith({ top: 0, behavior: "smooth" });
        });
        await waitFor(() => {
            expect(mockNavigateWithSearchParams).toHaveBeenCalledTimes(1);
        });
        expect(mockNavigateWithSearchParams).toHaveBeenCalledWith(`/${mockCreatedMovie.id}`);

        expect(screen.queryByTestId("mock-error-message")).not.toBeInTheDocument();
    });

    test("Should handle failed form submission", async () => {
        const mockFormData: MovieFormData = {
            title: "New Movie",
            poster_path: "path/to/image.jpg",
            genres: ["Action"],
            overview: "A great movie",
            runtime: 120,
            release_date: "2023-01-01",
            vote_average: 8.0,
        };
        const mockError = new Error("API Error");

        mockCreateMovie.mockImplementation(async () => {
            await delay(10);
            throw mockError;
        });
        const consoleErrorSpy = jest.spyOn(console, "error").mockImplementation(() => {});

        renderComponent();

        await act(async () => {
            mockMovieFormSubmit(mockFormData);
        });

        await waitFor(() => expect(lastMovieFormProps?.isLoading).toBe(true));

        expect(mockCreateMovie).toHaveBeenCalledTimes(1);
        expect(mockCreateMovie).toHaveBeenCalledWith(mockFormData);

        await waitFor(() => {
            expect(consoleErrorSpy).toHaveBeenCalledWith("Error creating a movie:", mockError);
        });
        await waitFor(() => {
            expect(screen.getByTestId("mock-error-message")).toBeInTheDocument();
        });
        expect(screen.getByTestId("mock-error-message")).toHaveTextContent(
            "Failed to create a movie. Please try again.",
        );

        expect(mockNavigateWithSearchParams).not.toHaveBeenCalled();
        expect(window.scrollTo).not.toHaveBeenCalled();

        consoleErrorSpy.mockRestore();
    });

    test("Should handle dialog close", async () => {
        renderComponent();

        await act(async () => {
            mockDialogOnClose();
        });

        await waitFor(() => {
            expect(mockNavigateWithSearchParams).toHaveBeenCalledTimes(1);
        });
        expect(mockNavigateWithSearchParams).toHaveBeenCalledWith("/");
    });

    test("Should handle error message dismissal", async () => {
        const mockError = new Error("API Error");
        mockCreateMovie.mockRejectedValue(mockError);
        jest.spyOn(console, "error").mockImplementation(() => {});

        renderComponent();

        await act(async () => {
            mockMovieFormSubmit({} as MovieFormData);
        });

        await waitFor(() => {
            expect(screen.getByTestId("mock-error-message")).toBeInTheDocument();
        });

        expect(mockErrorMessageDismiss).toBeDefined();

        await act(async () => {
            if (mockErrorMessageDismiss) {
                mockErrorMessageDismiss();
            }
        });

        await waitFor(() => {
            expect(screen.queryByTestId("mock-error-message")).not.toBeInTheDocument();
        });

        jest.restoreAllMocks();
    });
});
