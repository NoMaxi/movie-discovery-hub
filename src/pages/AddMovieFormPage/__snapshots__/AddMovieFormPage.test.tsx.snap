// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMovieFormPage Should render correctly with Dialog and MovieForm 1`] = `
<DocumentFragment>
  <div
    class="max-w-[976px]"
    data-testid="mock-dialog"
  >
    <h2
      data-testid="mock-dialog-title"
    >
      Add Movie
    </h2>
    <button
      data-testid="mock-dialog-close-button"
    >
      Close
    </button>
    <form
      data-testid="mock-movie-form"
    >
      Mock Movie Form
    </form>
  </div>
</DocumentFragment>
`;
